Steps to deploy to Heroku

Hi!

I'd like to know which steps you guys take to deploy a project using cookiecutter-flask to Heroku. I had to manually install dev dependencies so `npm run build` could be executed which sounded kind of weird to me..

Any help/guidance is appreciated!

Thank you!

@gcrsaldanha were you able to deploy to Heroku successfully? I'm getting errors:

```

Starting process with command `gunicorn erdos.app:create_app\(\)`
[INFO] Starting gunicorn 19.7.1
[INFO] Listening at: http://0.0.0.0:16541 (4)
[INFO] Using worker: sync
[INFO] Booting worker with pid: 9
[INFO] Booting worker with pid: 10
[INFO] Booting worker with pid: 11
State changed from starting to up

[ERROR] Exception in worker process
Traceback (most recent call last):
  File "/app/.heroku/python/lib/python2.7/site-packages/gunicorn/arbiter.py", line 578, in spawn_worker
    worker.init_process()
  File "/app/.heroku/python/lib/python2.7/site-packages/gunicorn/workers/base.py", line 126, in init_process
    self.load_wsgi()
  File "/app/.heroku/python/lib/python2.7/site-packages/gunicorn/workers/base.py", line 135, in load_wsgi
    self.wsgi = self.app.wsgi()
  File "/app/.heroku/python/lib/python2.7/site-packages/gunicorn/app/base.py", line 67, in wsgi
    self.callable = self.load()
  File "/app/.heroku/python/lib/python2.7/site-packages/gunicorn/app/wsgiapp.py", line 65, in load
    return self.load_wsgiapp()
  File "/app/.heroku/python/lib/python2.7/site-packages/gunicorn/app/wsgiapp.py", line 52, in load_wsgiapp
    return util.import_app(self.app_uri)
  File "/app/.heroku/python/lib/python2.7/site-packages/gunicorn/util.py", line 364, in import_app
    app = eval(obj, vars(mod))
  File "<string>", line 1, in <module>
  File "/app/erdos/app.py", line 19, in create_app
    register_extensions(app)
  File "/app/erdos/app.py", line 36, in register_extensions
    webpack.init_app(app)
  File "/app/.heroku/python/lib/python2.7/site-packages/flask_webpack/__init__.py", line 27, in init_app
    self._set_asset_paths(app)
  File "/app/.heroku/python/lib/python2.7/site-packages/flask_webpack/__init__.py", line 69, in _set_asset_paths
    "Flask-Webpack requires 'WEBPACK_MANIFEST_PATH' to be set and "
```

--- a/cookiecutter.json
+++ b/cookiecutter.json
@@ -7,5 +7,6 @@
 	"project_short_description": "A flasky app.",
 	"use_pipenv": ["no", "yes"],
 	"python_version": ["3.7", "3.6"],
-	"node_version": ["12", "10", "8"]
+	"node_version": ["12", "10", "8"],
+	"deployment_on_heroku": ["no", "yes"]
 }

--- a/cookiecutter.json
+++ b/cookiecutter.json
@@ -8,5 +8,5 @@
 	"use_pipenv": ["no", "yes"],
 	"python_version": ["3.7", "3.6"],
 	"node_version": ["12", "10", "8"],
-	"deployment_on_heroku": ["no", "yes"]
+	"use_heroku": ["no", "yes"]
 }

--- a/hooks/post_gen_project.py
+++ b/hooks/post_gen_project.py
@@ -10,12 +10,16 @@ import sys
 def clean_extra_package_management_files():
     """Removes either requirements files and folder or the Pipfile."""
     use_pipenv = "{{cookiecutter.use_pipenv}}"
+    use_heroku = "{{cookiecutter.use_heroku}}"
     to_delete = []
 
     if use_pipenv == "yes":
         to_delete = to_delete + ["requirements.txt", "requirements"]
     else:
         to_delete.append("Pipfile")
+
+    if use_heroku == "no":
+        to_delete = to_delete + ["Procfile", "app.json"]
 
     try:
         for file_or_dir in to_delete:

--- a/{{cookiecutter.app_name}}/.travis.yml
+++ b/{{cookiecutter.app_name}}/.travis.yml
@@ -12,6 +12,5 @@ install:
 before_script:
   - cp .env.example .env
   - npm run lint
-  - npm run build
   - flask lint --check
 script: flask test

--- a/{{cookiecutter.app_name}}/Pipfile
+++ b/{{cookiecutter.app_name}}/Pipfile
@@ -12,6 +12,9 @@ click = ">=5.0"
 # Database
 Flask-SQLAlchemy = "==2.4.1"
 SQLAlchemy = "==1.3.9"
+{%- if cookiecutter.deployment_on_heroku == "yes" %}
+psycopg2 = "==2.8.3"
+{%- endif %}
 
 # Migrations
 Flask-Migrate = "==2.5.2"

--- a/{{cookiecutter.app_name}}/Pipfile
+++ b/{{cookiecutter.app_name}}/Pipfile
@@ -12,8 +12,8 @@ click = ">=5.0"
 # Database
 Flask-SQLAlchemy = "==2.4.1"
 SQLAlchemy = "==1.3.9"
-{%- if cookiecutter.deployment_on_heroku == "yes" %}
-psycopg2 = "==2.8.3"
+{%- if cookiecutter.use_heroku == "yes" %}
+psycopg2-binary = "==2.8.3"
 {%- endif %}
 
 # Migrations

--- a/{{cookiecutter.app_name}}/Procfile
+++ b/{{cookiecutter.app_name}}/Procfile
@@ -1,1 +1,2 @@
+release: flask db upgrade
 web: gunicorn {{cookiecutter.app_name}}.app:create_app\(\) -b 0.0.0.0:$PORT -w 3

--- a/{{cookiecutter.app_name}}/README.rst
+++ b/{{cookiecutter.app_name}}/README.rst
@@ -85,7 +85,7 @@ To apply the migration.
 
 For a full migration command reference, run ``flask db --help``.
 
-If your deployment process is based on the git repository files (e.g on Heroku) you should add `migrations` folder to the repository.
+If you will deploy your application remotely (e.g on Heroku) you should add the `migrations` folder to version control.
 You can do this after ``flask db migrate`` by running the following commands ::
 
     git add migrations/*
@@ -149,38 +149,37 @@ in your ``settings.py``::
 Heroku
 ------
 
-Before deployment on the Heroku you should be familiar with a basic concepts of `Git <https://git-scm.com/>`_ and `Heroku <https://heroku.com/>`_.
+Before deploying to Heroku you should be familiar with the basic concepts of `Git <https://git-scm.com/>`_ and `Heroku <https://heroku.com/>`_.
 
 Remember to add migrations to your repository. Please check `Migrations`_ section.
 
-Since filesystem on Heroku is ephemeral (sqlite database will be lost at least once every 24 hours) the PostgreSQL is a recommended choice. Due to this fact the additional library is required: ``psycopg2-binary``.
+Since the filesystem on Heroku is ephemeral, non-version controlled files (like a SQLite database) will be lost at least once every 24 hours. Therefore, a persistent, standalone database like PostgreSQL is recommended. This application will work with any database backend that is compatible with SQLAlchemy, but we provide specific instructions for Postgres, (including the required library ``psycopg2-binary``).
 
-**WARNING!** ``psycopg2-binary`` package is a practical choice for development and testing but in production it is advised to use the package built from sources. Read more in the `psycopg2 documentation <http://initd.org/psycopg/docs/install.html?highlight=production%20advised%20use%20package%20built%20from%20sources#binary-install-from-pypi>`_
+**Note:** ``psycopg2-binary`` package is a practical choice for development and testing but in production it is advised to use the package built from sources. Read more in the `psycopg2 documentation <http://initd.org/psycopg/docs/install.html?highlight=production%20advised%20use%20package%20built%20from%20sources#binary-install-from-pypi>`_
 
 Deployment by using `Heroku CLI <https://devcenter.heroku.com/articles/heroku-cli>`_:
 
-* create Heroku App. You can leave your app name, change it or leave it blank (random name will be generated)::
+* Create Heroku App. You can leave your app name, change it, or leave it blank (random name will be generated)::
 
     heroku create {{cookiecutter.app_name}}
 
-* add buildpacks::
+* Add buildpacks::
 
     heroku buildpacks:add --index=1 heroku/nodejs
     heroku buildpacks:add --index=1 heroku/python
 
-* add database addon which sets PostgresSQL in version 11 in free `hobby-dev <https://elements.heroku.com/addons/heroku-postgresql#hobby-dev>`_ plan (it also sets ``DATABASE_URL`` environmental variable to created database)::
+* Add database addon which creates a persistent PostgresSQL database. These instructions assume you're using the free `hobby-dev <https://elements.heroku.com/addons/heroku-postgresql#hobby-dev>`_ plan. This command also sets a ``DATABASE_URL`` environmental variable that your app will use to communicate with the DB.::
 
     heroku addons:create heroku-postgresql:hobby-dev --version=11
 
-* set environmental variables (change ``SECRET_KEY`` value for your own)::
+* Set environmental variables (change ``SECRET_KEY`` value)::
 
     heroku config:set SECRET_KEY=not-so-secret
     heroku config:set FLASK_APP=autoapp.py
 
-*   ``DATABASE_URL`` environmental variable is set by database addon.
-    Please check ``.env.example`` to get overview which environmental variables are used in the project.
+*   Please check ``.env.example`` to see which environmental variables are used in the project and also need to be set. The exception is ``DATABASE_URL``, which Heroku sets automatically.
 
-* deploy on Heroku by pushing to the ``heroku`` branch::
+* Deploy on Heroku by pushing to the ``heroku`` branch::
 
     git push heroku master
 

--- a/{{cookiecutter.app_name}}/README.rst
+++ b/{{cookiecutter.app_name}}/README.rst
@@ -136,8 +136,8 @@ in your ``settings.py``::
 
     SEND_FILE_MAX_AGE_DEFAULT = 31556926  # one year
 
+{%- if cookiecutter.use_heroku == "yes" %}
 
-{%- if cookiecutter.deployment_on_heroku == "yes" %}
 Deployment on Heroku
 --------------------
 
@@ -174,13 +174,15 @@ If you want deploy by using Heroku CLI:
     heroku buildpacks:add --index=1 heroku/nodejs
     heroku buildpacks:add --index=1 heroku/python
 
-* add Postgres database addon (it also sets `DATABASE_URL` environmental variable to created database)::
+* add database addon which sets Postgres in version 11 in free `hobby-dev` plan (https://elements.heroku.com/addons/heroku-postgresql#hobby-dev) (it also sets `DATABASE_URL` environmental variable to created database)::
 
-    heroku addons:create heroku-postgresql:hobby-dev
+    heroku addons:create heroku-postgresql:hobby-dev --version=11
 
-* set environmental variables (change secret key)::
+* set environmental variables: change secret key.
+  `DATABASE_URL` environmental variable is set by database addon in previous step.
+  Please check `.env.example` to get overview which environmental variables are used in project.::
 
-    heroku config:set SECRET_KEY=<secret-key>
+    heroku config:set SECRET_KEY=not-so-secret
     heroku config:set FLASK_APP=autoapp.py
 
 * deploy on Heroku::

--- a/{{cookiecutter.app_name}}/README.rst
+++ b/{{cookiecutter.app_name}}/README.rst
@@ -85,6 +85,14 @@ To apply the migration.
 
 For a full migration command reference, run ``flask db --help``.
 
+If your deployment process is based on the git repository files (e.g on Heroku) you should add `migrations` folder to the repository.
+You can do this after ``flask db migrate`` by running the following commands ::
+
+    git add migrations/*
+    git commit -m "Add migrations"
+
+Make sure folder `migrations/versions` is not empty.
+
 
 Docker
 ------
@@ -138,32 +146,18 @@ in your ``settings.py``::
 
 {%- if cookiecutter.use_heroku == "yes" %}
 
-Deployment on Heroku
---------------------
+Heroku
+------
 
-Before using automatic deployment on Heroku you have to add migrations to your repository.
-You can do it by using following commands ::
+Before deployment on the Heroku you should be familiar with a basic concepts of `Git <https://git-scm.com/>`_ and `Heroku <https://heroku.com/>`_.
 
-    flask db init
-    flask db migrate
-    git add migrations/
-    git commit -m "Add migrations"
-    git commit push
+Remember to add migrations to your repository. Please check `Migrations`_ section.
 
-Make sure folder `migrations/versions` is not empty.
+Since filesystem on Heroku is ephemeral (sqlite database will be lost at least once every 24 hours) the PostgreSQL is a recommended choice. Due to this fact the additional library is required: ``psycopg2-binary``.
 
-Deploy to Heroku button
-^^^^^^^^^^^^^^^^^^^^^^^
+**WARNING!** ``psycopg2-binary`` package is a practical choice for development and testing but in production it is advised to use the package built from sources. Read more in the `psycopg2 documentation <http://initd.org/psycopg/docs/install.html?highlight=production%20advised%20use%20package%20built%20from%20sources#binary-install-from-pypi>`_
 
-.. raw:: html
-
-    <a href="https://heroku.com/deploy"><img src="https://www.herokucdn.com/deploy/button.svg" title="Deploy" alt="Deploy"></a>
-
-
-Heroku CLI
-^^^^^^^^^^
-
-If you want deploy by using Heroku CLI:
+Deployment by using `Heroku CLI <https://devcenter.heroku.com/articles/heroku-cli>`_:
 
 * create Heroku App. You can leave your app name, change it or leave it blank (random name will be generated)::
 
@@ -174,19 +168,27 @@ If you want deploy by using Heroku CLI:
     heroku buildpacks:add --index=1 heroku/nodejs
     heroku buildpacks:add --index=1 heroku/python
 
-* add database addon which sets Postgres in version 11 in free `hobby-dev` plan (https://elements.heroku.com/addons/heroku-postgresql#hobby-dev) (it also sets `DATABASE_URL` environmental variable to created database)::
+* add database addon which sets PostgresSQL in version 11 in free `hobby-dev <https://elements.heroku.com/addons/heroku-postgresql#hobby-dev>`_ plan (it also sets ``DATABASE_URL`` environmental variable to created database)::
 
     heroku addons:create heroku-postgresql:hobby-dev --version=11
 
-* set environmental variables: change secret key.
-  `DATABASE_URL` environmental variable is set by database addon in previous step.
-  Please check `.env.example` to get overview which environmental variables are used in project.::
+* set environmental variables (change ``SECRET_KEY`` value for your own)::
 
     heroku config:set SECRET_KEY=not-so-secret
     heroku config:set FLASK_APP=autoapp.py
 
-* deploy on Heroku::
+*   ``DATABASE_URL`` environmental variable is set by database addon.
+    Please check ``.env.example`` to get overview which environmental variables are used in the project.
+
+* deploy on Heroku by pushing to the ``heroku`` branch::
 
     git push heroku master
+
+If you keep your project on GitHub you can use 'Deploy to Heroku' button thanks to which the deployment can be done in web browser with minimal configuration required.
+The configuration used by the button is stored in ``app.json`` file.
+
+.. raw:: html
+
+    <a href="https://heroku.com/deploy"><img src="https://www.herokucdn.com/deploy/button.svg" title="Deploy" alt="Deploy"></a>
 
 {%- endif %}

--- a/{{cookiecutter.app_name}}/README.rst
+++ b/{{cookiecutter.app_name}}/README.rst
@@ -157,6 +157,14 @@ Since the filesystem on Heroku is ephemeral, non-version controlled files (like 
 
 **Note:** ``psycopg2-binary`` package is a practical choice for development and testing but in production it is advised to use the package built from sources. Read more in the `psycopg2 documentation <http://initd.org/psycopg/docs/install.html?highlight=production%20advised%20use%20package%20built%20from%20sources#binary-install-from-pypi>`_
 
+If you keep your project on GitHub you can use 'Deploy to Heroku' button thanks to which the deployment can be done in web browser with minimal configuration required.
+The configuration used by the button is stored in ``app.json`` file.
+
+.. raw:: html
+
+    <a href="https://heroku.com/deploy" style="display: block"><img src="https://www.herokucdn.com/deploy/button.svg" title="Deploy" alt="Deploy"></a>
+    <br>
+
 Deployment by using `Heroku CLI <https://devcenter.heroku.com/articles/heroku-cli>`_:
 
 * Create Heroku App. You can leave your app name, change it, or leave it blank (random name will be generated)::
@@ -182,12 +190,5 @@ Deployment by using `Heroku CLI <https://devcenter.heroku.com/articles/heroku-cl
 * Deploy on Heroku by pushing to the ``heroku`` branch::
 
     git push heroku master
-
-If you keep your project on GitHub you can use 'Deploy to Heroku' button thanks to which the deployment can be done in web browser with minimal configuration required.
-The configuration used by the button is stored in ``app.json`` file.
-
-.. raw:: html
-
-    <a href="https://heroku.com/deploy"><img src="https://www.herokucdn.com/deploy/button.svg" title="Deploy" alt="Deploy"></a>
 
 {%- endif %}

--- a/{{cookiecutter.app_name}}/README.rst
+++ b/{{cookiecutter.app_name}}/README.rst
@@ -135,3 +135,56 @@ should cache all your assets forever by including the following line
 in your ``settings.py``::
 
     SEND_FILE_MAX_AGE_DEFAULT = 31556926  # one year
+
+
+{%- if cookiecutter.deployment_on_heroku == "yes" %}
+Deployment on Heroku
+--------------------
+
+Before using automatic deployment on Heroku you have to add migrations to your repository.
+You can do it by using following commands ::
+
+    flask db init
+    flask db migrate
+    git add migrations/
+    git commit -m "Add migrations"
+    git commit push
+
+Make sure folder `migrations/versions` is not empty.
+
+Deploy to Heroku button
+^^^^^^^^^^^^^^^^^^^^^^^
+
+.. raw:: html
+
+    <a href="https://heroku.com/deploy"><img src="https://www.herokucdn.com/deploy/button.svg" title="Deploy" alt="Deploy"></a>
+
+
+Heroku CLI
+^^^^^^^^^^
+
+If you want deploy by using Heroku CLI:
+
+* create Heroku App. You can leave your app name, change it or leave it blank (random name will be generated)::
+
+    heroku create {{cookiecutter.app_name}}
+
+* add buildpacks::
+
+    heroku buildpacks:add --index=1 heroku/nodejs
+    heroku buildpacks:add --index=1 heroku/python
+
+* add Postgres database addon (it also sets `DATABASE_URL` environmental variable to created database)::
+
+    heroku addons:create heroku-postgresql:hobby-dev
+
+* set environmental variables (change secret key)::
+
+    heroku config:set SECRET_KEY=<secret-key>
+    heroku config:set FLASK_APP=autoapp.py
+
+* deploy on Heroku::
+
+    git push heroku master
+
+{%- endif %}

--- a/{{cookiecutter.app_name}}/app.json
+++ b/{{cookiecutter.app_name}}/app.json
@@ -1,6 +1,5 @@
 {
    "name": "{{cookiecutter.app_name}}",
-   "repository": "https://github.com/{{cookiecutter.github_username}}/{{cookiecutter.app_name}}",
    "env": {
       "SECRET_KEY": {
          "description": "SECRET_KEY.",

--- a/{{cookiecutter.app_name}}/app.json
+++ b/{{cookiecutter.app_name}}/app.json
@@ -1,0 +1,30 @@
+{
+   "name": "{{cookiecutter.app_name}}",
+   "repository": "https://github.com/{{cookiecutter.github_username}}/{{cookiecutter.app_name}}",
+   "env": {
+      "SECRET_KEY": {
+         "description": "SECRET_KEY.",
+         "generator": "secret"
+      },
+      "FLASK_APP": {
+         "description": "FLASK_APP.",
+         "value": "autoapp.py"
+      }
+   },
+   "buildpacks": [
+      {
+         "url": "heroku/nodejs"
+      },
+      {
+         "url": "heroku/python"
+      }
+   ],
+   "addons": [
+      {
+         "plan": "heroku-postgresql",
+         "options": {
+            "version": "11"
+         }
+      }
+   ]
+}

--- a/{{cookiecutter.app_name}}/app.json
+++ b/{{cookiecutter.app_name}}/app.json
@@ -21,7 +21,7 @@
    ],
    "addons": [
       {
-         "plan": "heroku-postgresql",
+         "plan": "heroku-postgresql:hobby-dev",
          "options": {
             "version": "11"
          }

--- a/{{cookiecutter.app_name}}/package.json
+++ b/{{cookiecutter.app_name}}/package.json
@@ -7,7 +7,8 @@
     "start": "concurrently -n \"WEBPACK,FLASK\" -c \"bgBlue.bold,bgMagenta.bold\" \"npm run webpack-dev-server\" \"npm run flask-server\"",
     "webpack-dev-server": "NODE_ENV=debug webpack-dev-server --host=0.0.0.0 --port 2992 --hot --inline",
     "flask-server": "{% if cookiecutter.use_pipenv == 'yes' %}pipenv run {% endif %}flask run --host=0.0.0.0",
-    "lint": "eslint \"assets/js/*.js\""
+    "lint": "eslint \"assets/js/*.js\"",
+    "postinstall": "npm run build"
   },
   "repository": {
     "type": "git",

--- a/{{cookiecutter.app_name}}/requirements/prod.txt
+++ b/{{cookiecutter.app_name}}/requirements/prod.txt
@@ -8,8 +8,8 @@ click>=7.0
 # Database
 Flask-SQLAlchemy==2.4.1
 SQLAlchemy==1.3.9
-{%- if cookiecutter.deployment_on_heroku == "yes" %}
-psycopg2==2.8.3
+{%- if cookiecutter.use_heroku == "yes" %}
+psycopg2-binary==2.8.3
 {%- endif %}
 
 # Migrations

--- a/{{cookiecutter.app_name}}/requirements/prod.txt
+++ b/{{cookiecutter.app_name}}/requirements/prod.txt
@@ -8,6 +8,9 @@ click>=7.0
 # Database
 Flask-SQLAlchemy==2.4.1
 SQLAlchemy==1.3.9
+{%- if cookiecutter.deployment_on_heroku == "yes" %}
+psycopg2==2.8.3
+{%- endif %}
 
 # Migrations
 Flask-Migrate==2.5.2

